src/drivers/net/dwc_gmac/dwc_gmac.c

```diff
@@ -0,0 +1,596 @@
+	phy_autoneg(dev, 0);
+}
+
+static void dwc_setup_rxfilter(struct dwc_priv *dwc_priv) {
+	uint8_t *hwaddr;
+	int ffval;
+	int lo;
+	int hi;
+
+	ffval = 0;
+	hwaddr = dwc_priv->macaddr;
+
+	/*
+	 * Set the multicast (group) filter hash.
+	 */
+	ffval = (DWC_GMAC_FRAME_FILTER_PM);
+
+	/*
+	 * Set the individual address filter hash.
+	 */
+	ffval |= (DWC_GMAC_FRAME_FILTER_PR);
+
+	/*
+	 * Set the primary address.
+	 */
+	lo = hwaddr[0] | (hwaddr[1] << 8) | (hwaddr[2] << 16) |
+	    (hwaddr[3] << 24);
+	hi = hwaddr[4] | (hwaddr[5] << 8);
+	dwc_reg_write(dwc_priv, DWC_GMAC_ADDRESS_LOW(0), lo);
+	dwc_reg_write(dwc_priv, DWC_GMAC_ADDRESS_HIGH(0), hi);
+	dwc_reg_write(dwc_priv, DWC_GMAC_FRAME_FILTER, ffval);
+
+	log_debug("hwaddr = %x:%x:%x:%x:%x:%x", hwaddr[0], hwaddr[1], hwaddr[2],
+			hwaddr[3], hwaddr[4], hwaddr[5]);
+}
+
+static void dwc_gmac_start(struct dwc_priv *dwc_priv) {
+	uint32_t reg;
+
+	dwc_setup_rxfilter(dwc_priv);
+
+	/* Initialize DMA and enable transmitters */
+	reg = dwc_reg_read(dwc_priv, DWC_DMA_OPERATION_MODE);
+	reg |= (DWC_DMA_MODE_TSF | DWC_DMA_MODE_OSF | DWC_DMA_MODE_FUF);
+	reg &= ~(DWC_DMA_MODE_RSF);
+	reg |= (DWC_DMA_MODE_RTC_LEV32 << DWC_DMA_MODE_RTC_SHIFT);
+	dwc_reg_write(dwc_priv, DWC_DMA_OPERATION_MODE, reg);
+
+	dwc_reg_write(dwc_priv, DWC_DMA_INTERRUPT_ENABLE, DWC_DMA_INT_EN_DEFAULT);
+
+	/* Start DMA */
+	reg = dwc_reg_read(dwc_priv, DWC_DMA_OPERATION_MODE);
+	reg |= (DWC_DMA_MODE_ST | DWC_DMA_MODE_SR);
+	dwc_reg_write(dwc_priv, DWC_DMA_OPERATION_MODE, reg);
+
+	/* Enable transmitters */
+	reg = dwc_reg_read(dwc_priv, DWC_GMAC_CONFIG);
+	reg |= (DWC_GMAC_CONF_JD | DWC_GMAC_CONF_ACS | DWC_GMAC_CONF_BE);
+	reg |= (DWC_GMAC_CONF_PS | DWC_GMAC_CONF_FES | DWC_GMAC_CONF_DM); /* 100mb mii */
+	reg |= (DWC_GMAC_CONF_TE | DWC_GMAC_CONF_RE);
+	dwc_reg_write(dwc_priv, DWC_GMAC_CONFIG, reg);
+}
+
+static int dwc_open(struct net_device *dev) {
+	uint32_t version;
+	struct dwc_priv *dwc_priv;
+
+	assert(dev);
+
+	dwc_priv = dev->priv;
+
+	version = dwc_reg_read(dwc_priv, DWC_GMAC_VERSION);
+	dwc_priv->mii_clk = (dwc_reg_read(dwc_priv, DWC_GMAC_MII_ADDR) >>
+			DWC_GMAC_GMII_ADDRESS_CR_SHIFT) & DWC_GMAC_GMII_ADDRESS_CR_MASK;
+
+	log_debug("version = %#X mmiclk %#X", version, dwc_priv->mii_clk);
+
+	dwc_mdio_init(dev);
+
+	dwc_gmac_start(dwc_priv);
+
+	return 0;
+}
+
+static int dwc_setup_txdesc(int idx, uint32_t buff, int len) {
+	struct dma_extended_desc *desc;
+	int next_desc;
+
+	assert(idx < TX_DESC_QUANTITY);
+
+	desc = &tx_descs_pool[idx];
+
+	desc->basic.des2 = buff;
+	desc->basic.des1 = len & 0x1FFF;
+
+	dcache_flush((void *)desc->basic.des2, len);
+
+	desc->basic.des0 = \
+			ETDES0_SECOND_ADDRESS_CHAINED | \
+			ETDES0_FIRST_SEGMENT | \
+			ETDES0_LAST_SEGMENT | \
+			ETDES0_INTERRUPT;
+	if (idx < (TX_DESC_QUANTITY - 1)) {
+		next_desc = idx +1;
+	} else {
+		next_desc = 0;
+		desc->basic.des0 |= ETDES0_END_RING;
+	}
+
+	desc->basic.des3 = (uint32_t)&tx_descs_pool[next_desc];
+	dcache_flush(desc, sizeof(*desc));
```

Comment:
Is this flush necessary? Another one is called two lines below

Reply:
Yes. I think so, DMA must see OWN bit strongly after it saw another field of descriptor. It can be memory barriers, but for more assurance I set flush